File: TheStyk-Windows/TheStyk-Windows/Models/NoteStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using TheStyk.Models;

namespace TheStyk.Models
{
    public class NoteStore
    {
        private static readonly Lazy<NoteStore> _instance = new(() => new NoteStore());
        public static NoteStore Shared => _instance.Value;

        public static readonly string IndexDidChangeNotification = "thestyk.indexDidChange";
        public event EventHandler? IndexChanged;

        private const long MaxNoteBytes = 2 * 1024 * 1024; // 2 MB
        private const long MaxIndexBytes = 5 * 1024 * 1024; // 5 MB
        private const int MaxEntries = 10000;
        private const int MaxTextChars = 200000;
        private static readonly TimeSpan TrashRetention = TimeSpan.FromDays(5);

        private readonly string _dataDir;
        private readonly string _notesDir;
        private readonly string _trashDir;
        private readonly string _indexFile;

        private readonly List<IndexEntry> _index = new();
        private readonly List<TrashEntry> _trash = new();
        private DateTime _lastReconcile = DateTime.MinValue;

        private readonly JsonSerializerOptions _jsonOptions;

        public IReadOnlyList<IndexEntry> Index => _index.AsReadOnly();
        public IReadOnlyList<TrashEntry> Trash => _trash.AsReadOnly();

        private NoteStore()
        {
            // Busca diretório personalizado por Env (útil para testes isolados)
            string? envDir = Environment.GetEnvironmentVariable("THESTYK_DATA_DIR");
            if (!string.IsNullOrEmpty(envDir))
            {
                _dataDir = envDir;
            }
            else
            {
                _dataDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "The Styk");
            }

            _notesDir = Path.Combine(_dataDir, "notes");
            _trashDir = Path.Combine(_dataDir, "trash");
            _indexFile = Path.Combine(_dataDir, "index.json");

            Directory.CreateDirectory(_dataDir);
            Directory.CreateDirectory(_notesDir);
            Directory.CreateDirectory(_trashDir);

            _jsonOptions = new JsonSerializerOptions
            {
                WriteIndented = true,
                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
            };

            ReloadFromDisk();
            PurgeExpiredTrash();
        }

        public void ReloadFromDisk()
        {
            _index.Clear();
            _trash.Clear();

            var loaded = LoadIndexFile(_indexFile);
            _index.AddRange(loaded.Index);
            _trash.AddRange(loaded.Trash);

            PostIndexDidChange();
        }

        public static string NormalizePath(string path)
        {
            if (string.IsNullOrEmpty(path)) return string.Empty;
            
            // Substitui contra-barras por barras estilo POSIX para compatibilidade no index
            string normalized = path.Replace("\\", "/");

            // No Windows, caminhos de rede podem começar com "//" (UNC).
            // Remove barra final de caminhos normais (ex: "C:/Users/" vira "C:/Users", mas "C:/" permanece "C:/")
            if (normalized.EndsWith("/") && normalized.Length > 3)
            {
                normalized = normalized.TrimEnd('/');
            }

            return normalized;
        }

        // MARK: - Consulta

        public List<IndexEntry> Entries(string folder)
        {
            string normalized = NormalizePath(folder);
            // Comparação case-insensitive para caminhos do Windows
            return _index.Where(e => string.Equals(e.Folder, normalized, StringComparison.OrdinalIgnoreCase)).ToList();
        }

        public List<string> Folders()
        {
            return _index
                .Where(e => e.Orphaned != true)
                .Select(e => e.Folder)
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .OrderBy(f => f)
                .ToList();
        }

        public List<IndexEntry> Orphans()
        {
            return _index.Where(e => e.Orphaned == true).ToList();
        }

        // MARK: - Lazy load da nota completa

        public Note? LoadNote(Guid id)
        {
            string path = GetNotePath(id);
            if (!File.Exists(path))
            {
                SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: Arquivo nao existe em '{path}'");
                return null;
            }

            try
            {
                var fileInfo = new FileInfo(path);
                if (fileInfo.Length > MaxNoteBytes)
                {
                    SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: Tamanho do arquivo {fileInfo.Length} excede o limite {MaxNoteBytes}");
                    return null;
                }

                byte[] data = File.ReadAllBytes(path);
                if (data.Length > MaxNoteBytes)
                {
                    SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: Tamanho dos bytes {data.Length} excede o limite {MaxNoteBytes}");
                    return null;
                }

                // Pula o UTF-8 BOM se presente (0xEF, 0xBB, 0xBF) para evitar erro no JsonSerializer
                ReadOnlySpan<byte> span = data;
                if (span.Length >= 3 && span[0] == 0xEF && span[1] == 0xBB && span[2] == 0xBF)
                {
                    span = span.Slice(3);
                }

                var note = JsonSerializer.Deserialize<Note>(span, _jsonOptions);
                if (note == null)
                {
                    SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: Deserializacao JSON retornou null");
                    return null;
                }
                if (note.Id != id)
                {
                    SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: ID da nota ({note.Id}) nao bate com o ID buscado ({id})");
                    return null;
                }

                // Validações de guardrail
                if (string.IsNullOrEmpty(note.Folder))
                {
                    SystemIntegration.ExplorerObserver.Log($"[NoteStore] LoadNote falhou: Folder da nota esta vazio ou nulo");
                    return null;
                }

                note.Folder = NormalizePath(note.Folder);
                if (note.Text.Length > MaxTextChars)
                {
                    note.Text = note.Text.Substring(0, MaxTextChars);
                }

                note.Style = new NoteStyle
                {
                    Color = note.Style.Color,
                    FontID = note.Style.FontID,
                    FontSize = Clamp(note.Style.FontSize, 8, 72)
                };

                note.Frame = new NoteFrame
                {
                    X = double.IsFinite(note.Frame.X) ? note.Frame.X : 0,
                    Y = double.IsFinite(note.Frame.Y) ? note.Frame.Y : 0,
                    W = Clamp(note.Frame.W, 120, 4000),
                    H = Clamp(note.Frame.H, 120, 4000)
                };

                return note;
            }
            catch (Exception ex)
            {
                SystemIntegration.ExplorerObserver.Log($"[NoteStore] Erro ao carregar nota {id}: {ex.Message}\n{ex.StackTrace}");
                return null;
            }
        }

        // MARK: - Mutação

        public Note CreateNote(string folder, NoteFrame frame)
        {
            var now = DateTime.UtcNow;
            var note = new Note
            {
                Id = Guid.NewGuid(),
                Folder = NormalizePath(folder),
                Text = string.Empty,
                Style = new NoteStyle { Color = NoteColor.yellow, FontID = NoteFontID.system, FontSize = 14 },
                Frame = frame,
                Created = now,
                Modified = now
            };

            WriteNoteFile(note);

            _index.Add(CreateIndexEntry(note));
            WriteIndex();
            PostIndexDidChange();

            return note;
        }

        public void Save(Note note)
        {
            note.Modified = DateTime.UtcNow;
            note.Folder = NormalizePath(note.Folder);

            if (note.Text.Length > MaxTextChars)
            {
                note.Text = note.Text.Substring(0, MaxTextChars);
            }

            WriteNoteFile(note);

            var entry = CreateIndexEntry(note);
            int idx = _index.FindIndex(e => e.Id == note.Id);
            if (idx >= 0)
            {
                _index[idx] = entry;
            }
            else
            {
                _index.Add(entry);
            }

            WriteIndex();
            PostIndexDidChange();
        }

        // MARK: - Lixeira (recuperável por 5 dias)

        public void MoveToTrash(Guid id)
        {
            int idx = _index.FindIndex(e => e.Id == id);
            if (idx < 0) return;

            var entry = _index[idx];
            _index.RemoveAt(idx);

            string sourcePath = GetNotePath(id);
            string destPath = Path.Combine(_trashDir, id + ".json");

            try
            {
                if (File.Exists(destPath)) File.Delete(destPath);
                if (File.Exists(sourcePath))
                {
                    File.Move(sourcePath, destPath);
                    _trash.Add(new TrashEntry
                    {
                        Id = entry.Id,
                        Folder = entry.Folder,
                        Snippet = entry.Snippet,
                        Color = entry.Color,
                        DeletedAt = DateTime.UtcNow
                    });
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"The Styk: falha ao mover nota {id} para lixeira: {ex.Message}");
                if (File.Exists(sourcePath)) File.Delete(sourcePath);
            }

            WriteIndex();
            PostIndexDidChange();
        }

        public void RestoreFromTrash(Guid id)
        {
            int idx = _trash.FindIndex(t => t.Id == id);
            if (idx < 0) return;

            var trashEntry = _trash[idx];
            string sourcePath = Path.Combine(_trashDir, id + ".json");
            string destPath = GetNotePath(id);

            try
            {
                if (File.Exists(sourcePath))
                {
                    if (File.Exists(destPath)) File.Delete(destPath);
                    File.Move(sourcePath, destPath);
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"The Styk: falha ao restaurar nota {id}: {ex.Message}");
                return;
            }

            _trash.RemoveAt(idx);

            // Verifica se a pasta original existe. Se não existir, vira órfã.
            bool folderExists = Directory.Exists(trashEntry.Folder);

            _index.Add(new IndexEntry
            {
                Id = trashEntry.Id,
                Folder = trashEntry.Folder,
                Snippet = trashEntry.Snippet,
                Color = trashEntry.Color,
                Modified = DateTime.UtcNow,
                Orphaned = folderExists ? null : true
            });

            WriteIndex();
            PostIndexDidChange();
        }

        public void DeletePermanently(Guid id)
        {
            _trash.RemoveAll(t => t.Id == id);
            string trashPath = Path.Combine(_trashDir, id + ".json");
            try
            {
                if (File.Exists(trashPath)) File.Delete(trashPath);
            }
            catch {}

            WriteIndex();
            PostIndexDidChange();
        }

        public void EmptyTrash()
        {
            foreach (var t in _trash)
            {
                string path = Path.Combine(_trashDir, t.Id + ".json");
                try { if (File.Exists(path)) File.Delete(path); } catch {}
            }
            _trash.Clear();
            WriteIndex();
            PostIndexDidChange();
        }

        private void PurgeExpiredTrash()
        {
            DateTime cutoff = DateTime.UtcNow - TrashRetention;
            var expired = _trash.Where(t => t.DeletedAt < cutoff).ToList();
            if (expired.Count == 0) return;

            foreach (var t in expired)
            {
                string path = Path.Combine(_trashDir, t.Id + ".json");
                try { if (File.Exists(path)) File.Delete(path); } catch {}
            }

            _trash.RemoveAll(t => t.DeletedAt < cutoff);
            WriteIndex();
            PostIndexDidChange();
        }

        // MARK: - Âncoras: pasta apagada vira órfã, recriada des-órfã

        public void Reattach(Guid id, string toFolder)
        {
            var note = LoadNote(id);
            if (note == null) return;

            string normalizedFolder = NormalizePath(toFolder);
            note.Folder = normalizedFolder;
            note.Modified = DateTime.UtcNow;

            WriteNoteFile(note);

            int idx = _index.FindIndex(e => e.Id == id);
            if (idx >= 0)
            {
                _index[idx].Folder = normalizedFolder;
                _index[idx].Orphaned = null;
                _index[idx].Modified = note.Modified;
            }

            WriteIndex();
            PostIndexDidChange();
        }

        public void ReconcileAnchorsIfStale()
        {
            if ((DateTime.UtcNow - _lastReconcile).TotalSeconds > 30)
            {
                ReconcileAnchors();
            }
        }

        public void ReconcileAnchors()
        {
            _lastReconcile = DateTime.UtcNow;
            bool changed = false;

            // Agrupa por pasta única
            var folders = _index.Select(e => e.Folder).Distinct().ToList();

            foreach (var folder in folders)
            {
                bool exists = Directory.Exists(folder);
                bool isOrphan = _index.Any(e => string.Equals(e.Folder, folder, StringComparison.OrdinalIgnoreCase) && e.Orphaned == true);

                if (exists)
                {
                    if (isOrphan)
                    {
                        // Pasta voltou a existir -> deixa de ser órfã
                        foreach (var entry in _index.Where(e => string.Equals(e.Folder, folder, StringComparison.OrdinalIgnoreCase)))
                        {
                            entry.Orphaned = null;
                        }
                        changed = true;
                    }
                }
                else
                {
                    if (!isOrphan)
                    {
                        // Pasta sumiu -> vira órfã
                        foreach (var entry in _index.Where(e => string.Equals(e.Folder, folder, StringComparison.OrdinalIgnoreCase)))
                        {
                            entry.Orphaned = true;
                        }
                        changed = true;
                    }
                }
            }

            if (changed)
            {
                WriteIndex();
                PostIndexDidChange();
            }
        }

        // MARK: - Disco

        private string GetNotePath(Guid id) => Path.Combine(_notesDir, id + ".json");

        private void WriteNoteFile(Note note)
        {
            try
            {
                string path = GetNotePath(note.Id);
                string json = JsonSerializer.Serialize(note, _jsonOptions);
                WriteFileAtomic(path, json);
            }
            catch (Exception ex)
            {
                SystemIntegration.ExplorerObserver.Log($"[NoteStore] Falha ao gravar nota {note.Id}: {ex.Message}\n{ex.StackTrace}");
            }
        }

        private void WriteIndex()
        {
            try
            {
                var root = new
                {
                    version = 1,
                    notes = _index,
                    trash = _trash
                };
                string json = JsonSerializer.Serialize(root, _jsonOptions);
                WriteFileAtomic(_indexFile, json);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"The Styk: falha ao gravar index.json: {ex.Message}");
            }
        }

        private static void WriteFileAtomic(string filePath, string content)
        {
            string tempPath = filePath + ".tmp";
            // Usa UTF-8 sem BOM (Byte Order Mark) para compatibilidade com JsonSerializer.Deserialize
            var encoding = new System.Text.UTF8Encoding(false);
            File.WriteAllText(tempPath, content, encoding);

            try
            {
                if (File.Exists(filePath))
                {
                    File.Replace(tempPath, filePath, null);
                }
                else
                {
                    File.Move(tempPath, filePath);
                }
            }
            catch
            {
                // Fallback caso File.Replace fale devido a restrições/travas temporárias
                File.Copy(tempPath, filePath, true);
                File.Delete(tempPath);
            }
        }

        private static (List<IndexEntry> Index, List<TrashEntry> Trash) LoadIndexFile(string path)
        {
            var entries = new List<IndexEntry>();
            var trashed = new List<TrashEntry>();

            if (!File.Exists(path)) return (entries, trashed);

            try
            {
                var fileInfo = new FileInfo(path);
                if (fileInfo.Length > MaxIndexBytes) return (entries, trashed);

                byte[] data = File.ReadAllBytes(path);
                if (data.Length > MaxIndexBytes) return (entries, trashed); // TOCTOU guard

                using (var doc = JsonDocument.Parse(data))
                {
                    var root = doc.RootElement;
                    
                    if (root.TryGetProperty("notes", out var notesProp) && notesProp.ValueKind == JsonValueKind.Array)
                    {
                        foreach (var item in notesProp.EnumerateArray())
                        {
                            if (entries.Count >= MaxEntries) break;
                            try
                            {
                                var entry = JsonSerializer.Deserialize<IndexEntry>(item.GetRawText());
                                if (entry != null && !string.IsNullOrEmpty(entry.Folder))
                                {
                                    entry.Folder = NormalizePath(entry.Folder);
                                    entry.Snippet = MakeSnippet(entry.Snippet);
                                    entries.Add(entry);
                                }
                            }
                            catch { /* Descarta entrada malformada silenciando */ }
                        }
                    }

                    if (root.TryGetProperty("trash", out var trashProp) && trashProp.ValueKind == JsonValueKind.Array)
                    {
                        foreach (var item in trashProp.EnumerateArray())
                        {
                            if (trashed.Count >= MaxEntries) break;
                            try
                            {
                                var entry = JsonSerializer.Deserialize<TrashEntry>(item.GetRawText());
                                if (entry != null && !string.IsNullOrEmpty(entry.Folder))
                                {
                                    entry.Folder = NormalizePath(entry.Folder);
                                    entry.Snippet = MakeSnippet(entry.Snippet);
                                    trashed.Add(entry);
                                }
                            }
                            catch { /* Descarta entrada malformada silenciando */ }
                        }
                    }
                }
            }
            catch
            {
                // Descarta, nunca crasha
            }

            return (entries, trashed);
        }

        // MARK: - Helpers

        private IndexEntry CreateIndexEntry(Note note)
        {
            return new IndexEntry
            {
                Id = note.Id,
                Folder = note.Folder,
                Snippet = MakeSnippet(note.Text),
                Color = note.Style.Color,
                Modified = note.Modified
            };
        }

        public static string MakeSnippet(string text)
        {
            if (string.IsNullOrEmpty(text)) return "Nota vazia";

            var lines = text.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
            foreach (var line in lines)
            {
                var cleanChars = new System.Text.StringBuilder();
                foreach (char c in line)
                {
                    var cat = char.GetUnicodeCategory(c);
                    if (cat != System.Globalization.UnicodeCategory.Control &&
                        cat != System.Globalization.UnicodeCategory.Format)
                    {
                        cleanChars.Append(c);
                    }
                }
                string clean = cleanChars.ToString().Trim();
                if (!string.IsNullOrEmpty(clean))
                {
                    return clean.Length > 60 ? clean.Substring(0, 60) : clean;
                }
            }

            return "Nota vazia";
        }

        private static double Clamp(double v, double lo, double hi)
        {
            if (!double.IsFinite(v)) return lo;
            return Math.Min(Math.Max(v, lo), hi);
        }

        private void PostIndexDidChange()
        {
            // Notifica na thread principal se possível (ou dispara o event delegate)
            var context = SynchronizationContext.Current;
            if (context != null)
            {
                context.Post(_ => IndexChanged?.Invoke(this, EventArgs.Empty), null);
            }
            else
            {
                IndexChanged?.Invoke(this, EventArgs.Empty);
            }
        }
    }
}

File: TheStyk-Linux/src/note_store.py
import os
import json
import uuid
import shutil
from datetime import datetime, timedelta
from typing import List, Dict, Optional, Callable
from models import Note, NoteStyle, NoteFrame, IndexEntry, TrashEntry, NoteColor, NoteFontID, parse_iso_datetime, format_iso_datetime

class NoteStore:
    _instance: Optional["NoteStore"] = None

    @classmethod
    def get_shared(cls) -> "NoteStore":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def __init__(self):
        self.index_changed_callbacks: List[Callable[[], None]] = []
        
        self.max_note_bytes = 2 * 1024 * 1024  # 2 MB
        self.max_index_bytes = 5 * 1024 * 1024  # 5 MB
        self.max_entries = 10000
        self.max_text_chars = 200000
        self.trash_retention = timedelta(days=5)

        # Custom env data dir (for tests)
        env_dir = os.environ.get("THESTYK_DATA_DIR")
        if env_dir:
            self._data_dir = env_dir
        else:
            self._data_dir = os.path.join(os.path.expanduser("~"), ".config", "the-styk")

        self._notes_dir = os.path.join(self._data_dir, "notes")
        self._trash_dir = os.path.join(self._data_dir, "trash")
        self._index_file = os.path.join(self._data_dir, "index.json")

        os.makedirs(self._data_dir, exist_ok=True)
        os.makedirs(self._notes_dir, exist_ok=True)
        os.makedirs(self._trash_dir, exist_ok=True)

        self._index: List[IndexEntry] = []
        self._trash: List[TrashEntry] = []
        self._last_reconcile: datetime = datetime.min

        self.reload_from_disk()
        self.purge_expired_trash()

    def register_callback(self, callback: Callable[[], None]):
        if callback not in self.index_changed_callbacks:
            self.index_changed_callbacks.append(callback)

    def unregister_callback(self, callback: Callable[[], None]):
        if callback in self.index_changed_callbacks:
            self.index_changed_callbacks.remove(callback)

    def post_index_did_change(self):
        for cb in self.index_changed_callbacks:
            try:
                cb()
            except Exception as e:
                print(f"[NoteStore] Error in callback: {e}")

    @property
    def index(self) -> List[IndexEntry]:
        return self._index

    @property
    def trash(self) -> List[TrashEntry]:
        return self._trash

    def reload_from_disk(self):
        self._index.clear()
        self._trash.clear()

        loaded_index, loaded_trash = self._load_index_file(self._index_file)
        self._index.extend(loaded_index)
        self._trash.extend(loaded_trash)
        self.post_index_did_change()

    @staticmethod
    def normalize_path(path: str) -> str:
        if not path:
            return ""
        # POSIX normalize
        normalized = path.replace("\\", "/")
        if normalized.endswith("/") and len(normalized) > 1:
            normalized = normalized.rstrip("/")
        return normalized

    # MARK: - Queries
    def entries(self, folder: str) -> List[IndexEntry]:
        normalized = self.normalize_path(folder)
        return [e for e in self._index if e.folder.lower() == normalized.lower()]

    def folders(self) -> List[str]:
        # Return unique non-orphaned folders, sorted alphabetically
        unique_folders = set(e.folder for e in self._index if e.orphaned is not True)
        return sorted(list(unique_folders), key=lambda x: x.lower())

    def orphans(self) -> List[IndexEntry]:
        return [e for e in self._index if e.orphaned is True]

    # MARK: - Note Lazy Load
    def load_note(self, note_id: uuid.UUID) -> Optional[Note]:
        path = self._get_note_path(note_id)
        if not os.path.exists(path):
            print(f"[NoteStore] LoadNote falhou: Arquivo nao existe em '{path}'")
            return None

        try:
            sz = os.path.getsize(path)
            if sz > self.max_note_bytes:
                print(f"[NoteStore] LoadNote falhou: Tamanho do arquivo {sz} excede o limite {self.max_note_bytes}")
                return None

            with open(path, "rb") as f:
                data = f.read()

            if len(data) > self.max_note_bytes:
                print(f"[NoteStore] LoadNote falhou: Tamanho dos bytes {len(data)} excede o limite {self.max_note_bytes}")
                return None

            # Handle UTF-8 BOM
            if data.startswith(b"\xef\xbb\xbf"):
                data = data[3:]

            d = json.loads(data.decode("utf-8"))
            if not d:
                print(f"[NoteStore] LoadNote falhou: Deserializacao JSON retornou null")
                return None

            note = Note.from_dict(d)
            if note.id != note_id:
                print(f"[NoteStore] LoadNote falhou: ID nao bate")
                return None

            # Guardrails
            if not note.folder:
                print(f"[NoteStore] LoadNote falhou: Folder esta vazio")
                return None

            note.folder = self.normalize_path(note.folder)
            if len(note.text) > self.max_text_chars:
                note.text = note.text[:self.max_text_chars]

            note.style.fontSize = max(8.0, min(float(note.style.fontSize), 72.0))
            note.frame.x = float(note.frame.x) if hasattr(note.frame, "x") and note.frame.x is not None else 0.0
            note.frame.y = float(note.frame.y) if hasattr(note.frame, "y") and note.frame.y is not None else 0.0
            note.frame.w = max(120.0, min(float(note.frame.w), 4000.0))
            note.frame.h = max(120.0, min(float(note.frame.h), 4000.0))

            return note
        except Exception as e:
            print(f"[NoteStore] Erro ao carregar nota {note_id}: {e}")
            return None

    # MARK: - Mutations
    def create_note(self, folder: str, frame: NoteFrame) -> Note:
        now = datetime.utcnow()
        note = Note(
            id=uuid.uuid4(),
            folder=self.normalize_path(folder),
            text="",
            style=NoteStyle(color=NoteColor.yellow, fontID=NoteFontID.system, fontSize=14.0),
            frame=frame,
            created=now,
            modified=now
        )
        self._write_note_file(note)
        self._index.append(self._create_index_entry(note))
        self._write_index()
        self.post_index_did_change()
        return note

    def save(self, note: Note):
        note.modified = datetime.utcnow()
        note.folder = self.normalize_path(note.folder)
        if len(note.text) > self.max_text_chars:
            note.text = note.text[:self.max_text_chars]

        self._write_note_file(note)

        entry = self._create_index_entry(note)
        idx = -1
        for i, e in enumerate(self._index):
            if e.id == note.id:
                idx = i
                break

        if idx >= 0:
            self._index[idx] = entry
        else:
            self._index.append(entry)

        self._write_index()
        self.post_index_did_change()

    # MARK: - Trash
    def move_to_trash(self, note_id: uuid.UUID):
        idx = -1
        for i, e in enumerate(self._index):
            if e.id == note_id:
                idx = i
                break
        if idx < 0:
            return

        entry = self._index[idx]
        del self._index[idx]

        source_path = self._get_note_path(note_id)
        dest_path = os.path.join(self._trash_dir, f"{note_id}.json")

        try:
            if os.path.exists(dest_path):
                os.remove(dest_path)
            if os.path.exists(source_path):
                shutil.move(source_path, dest_path)
                self._trash.append(TrashEntry(
                    id=entry.id,
                    folder=entry.folder,
                    snippet=entry.snippet,
                    color=entry.color,
                    deletedAt=datetime.utcnow()
                ))
        except Exception as ex:
            print(f"[NoteStore] Falha ao mover nota {note_id} para lixeira: {ex}")
            if os.path.exists(source_path):
                try:
                    os.remove(source_path)
                except Exception:
                    pass

        self._write_index()
        self.post_index_did_change()

    def restore_from_trash(self, note_id: uuid.UUID):
        idx = -1
        for i, t in enumerate(self._trash):
            if t.id == note_id:
                idx = i
                break
        if idx < 0:
            return

        trash_entry = self._trash[idx]
        source_path = os.path.join(self._trash_dir, f"{note_id}.json")
        dest_path = self._get_note_path(note_id)

        try:
            if os.path.exists(source_path):
                if os.path.exists(dest_path):
                    os.remove(dest_path)
                shutil.move(source_path, dest_path)
        except Exception as ex:
            print(f"[NoteStore] Falha ao restaurar nota {note_id}: {ex}")
            return

        del self._trash[idx]
        folder_exists = os.path.isdir(trash_entry.folder)

        self._index.append(IndexEntry(
            id=trash_entry.id,
            folder=trash_entry.folder,
            snippet=trash_entry.snippet,
            color=trash_entry.color,
            modified=datetime.utcnow(),
            orphaned=None if folder_exists else True
        ))

        self._write_index()
        self.post_index_did_change()

    def delete_permanently(self, note_id: uuid.UUID):
        self._trash = [t for t in self._trash if t.id != note_id]
        trash_path = os.path.join(self._trash_dir, f"{note_id}.json")
        try:
            if os.path.exists(trash_path):
                os.remove(trash_path)
        except Exception:
            pass

        self._write_index()
        self.post_index_did_change()

    def empty_trash(self):
        for t in self._trash:
            path = os.path.join(self._trash_dir, f"{t.id}.json")
            try:
                if os.path.exists(path):
                    os.remove(path)
            except Exception:
                pass
        self._trash.clear()
        self._write_index()
        self.post_index_did_change()

    def purge_expired_trash(self):
        cutoff = datetime.utcnow() - self.trash_retention
        expired = [t for t in self._trash if t.deletedAt < cutoff]
        if not expired:
            return

        for t in expired:
            path = os.path.join(self._trash_dir, f"{t.id}.json")
            try:
                if os.path.exists(path):
                    os.remove(path)
            except Exception:
                pass

        self._trash = [t for t in self._trash if t.deletedAt >= cutoff]
        self._write_index()
        self.post_index_did_change()

    # MARK: - Anchors
    def reattach(self, note_id: uuid.UUID, to_folder: str):
        note = self.load_note(note_id)
        if not note:
            return

        normalized = self.normalize_path(to_folder)
        note.folder = normalized
        note.modified = datetime.utcnow()

        self._write_note_file(note)

        for entry in self._index:
            if entry.id == note_id:
                entry.folder = normalized
                entry.orphaned = None
                entry.modified = note.modified
                break

        self._write_index()
        self.post_index_did_change()

    def reconcile_anchors_if_stale(self):
        if (datetime.utcnow() - self._last_reconcile).total_seconds() > 30:
            self.reconcile_anchors()

    def reconcile_anchors(self):
        self._last_reconcile = datetime.utcnow()
        changed = False

        unique_folders = set(e.folder for e in self._index)
        for folder in unique_folders:
            exists = os.path.isdir(folder)
            is_orphan = any(e.folder.lower() == folder.lower() and e.orphaned is True for e in self._index)

            if exists:
                if is_orphan:
                    for entry in self._index:
                        if entry.folder.lower() == folder.lower():
                            entry.orphaned = None
                    changed = True
            else:
                if not is_orphan:
                    for entry in self._index:
                        if entry.folder.lower() == folder.lower():
                            entry.orphaned = True
                    changed = True

        if changed:
            self._write_index()
            self.post_index_did_change()

    # MARK: - Disk Helpers
    def _get_note_path(self, note_id: uuid.UUID) -> str:
        return os.path.join(self._notes_dir, f"{note_id}.json")

    def _write_note_file(self, note: Note):
        try:
            path = self._get_note_path(note.id)
            json_str = json.dumps(note.to_dict(), indent=2, ensure_ascii=False)
            self._write_file_atomic(path, json_str)
        except Exception as e:
            print(f"[NoteStore] Falha ao gravar nota {note.id}: {e}")

    def _write_index(self):
        try:
            root = {
                "version": 1,
                "notes": [e.to_dict() for e in self._index],
                "trash": [t.to_dict() for t in self._trash]
            }
            json_str = json.dumps(root, indent=2, ensure_ascii=False)
            self._write_file_atomic(self._index_file, json_str)
        except Exception as e:
            print(f"[NoteStore] Falha ao gravar index.json: {e}")

    @staticmethod
    def _write_file_atomic(file_path: str, content: str):
        temp_path = file_path + ".tmp"
        with open(temp_path, "w", encoding="utf-8") as f:
            f.write(content)

        # Atomic replacement
        try:
            if os.path.exists(file_path):
                os.replace(temp_path, file_path)
            else:
                os.rename(temp_path, file_path)
        except Exception:
            shutil.copy(temp_path, file_path)
            try:
                os.remove(temp_path)
            except Exception:
                pass

    def _load_index_file(self, path: str) -> tuple:
        entries = []
        trashed = []

        if not os.path.exists(path):
            return entries, trashed

        try:
            sz = os.path.getsize(path)
            if sz > self.max_index_bytes:
                return entries, trashed

            with open(path, "rb") as f:
                data = f.read()

            if len(data) > self.max_index_bytes:
                return entries, trashed

            if data.startswith(b"\xef\xbb\xbf"):
                data = data[3:]

            root = json.loads(data.decode("utf-8"))
            if not root or not isinstance(root, dict):
                return entries, trashed

            if "notes" in root and isinstance(root["notes"], list):
                for item in root["notes"]:
                    if len(entries) >= self.max_entries:
                        break
                    try:
                        entry = IndexEntry.from_dict(item)
                        if entry.folder:
                            entry.folder = self.normalize_path(entry.folder)
                            entry.snippet = self.make_snippet(entry.snippet)
                            entries.append(entry)
                    except Exception:
                        pass

            if "trash" in root and isinstance(root["trash"], list):
                for item in root["trash"]:
                    if len(trashed) >= self.max_entries:
                        break
                    try:
                        entry = TrashEntry.from_dict(item)
                        if entry.folder:
                            entry.folder = self.normalize_path(entry.folder)
                            entry.snippet = self.make_snippet(entry.snippet)
                            trashed.append(entry)
                    except Exception:
                        pass

        except Exception:
            pass

        return entries, trashed

    def _create_index_entry(self, note: Note) -> IndexEntry:
        return IndexEntry(
            id=note.id,
            folder=note.folder,
            snippet=self.make_snippet(note.text),
            color=note.style.color,
            modified=note.modified
        )

    @staticmethod
    def make_snippet(text: str) -> str:
        if not text:
            return "Nota vazia"

        lines = text.splitlines()
        for line in lines:
            # Strip control characters
            clean = "".join(c for c in line if ord(c) >= 32 and ord(c) != 127).strip()
            if clean:
                return clean[:60] if len(clean) > 60 else clean

        return "Nota vazia"
